klippy/extras/axis_twist_compensation.py

```diff
@@ -142,16 +241,155 @@ def cmd_AXIS_TWIST_COMPENSATION_CALIBRATE(self, gcmd):
         # begin calibration
         self.current_point_index = 0
         self.results = []
+        self.current_axis = axis
         self._calibration(probe_points, nozzle_points, interval_dist)
 
-    def _calculate_nozzle_points(self, sample_count, interval_dist):
-        # calculate the points to put the probe at, returned as a list of tuples
-        nozzle_points = []
+    def _calculate_corrections(self, coordinates):
+        # Extracting x, y, and z values from coordinates
+        x_coords = [coord[0] for coord in coordinates]
+        y_coords = [coord[1] for coord in coordinates]
+        z_coords = [coord[2] for coord in coordinates]
+
+        # Calculate the desired point (average of all corner points in z)
+        # For a general case, we should extract the unique
+        # combinations of corner points
+        z_corners = [z_coords[i] for i, coord in enumerate(coordinates)
+                        if (coord[0] in [x_coords[0], x_coords[-1]])
+                        and (coord[1] in [y_coords[0], y_coords[-1]])]
+        z_desired = sum(z_corners) / len(z_corners)
+
+
+        # Calculate average deformation per axis
+        unique_x_coords = sorted(set(x_coords))
+        unique_y_coords = sorted(set(y_coords))
+
+        avg_z_x = []
+        for x in unique_x_coords:
+            indices = [i for i, coord in enumerate(coordinates)
+                        if coord[0] == x]
+            avg_z = sum(z_coords[i] for i in indices) / len(indices)
+            avg_z_x.append(avg_z)
+
+        avg_z_y = []
+        for y in unique_y_coords:
+            indices = [i for i, coord in enumerate(coordinates)
+                        if coord[1] == y]
+            avg_z = sum(z_coords[i] for i in indices) / len(indices)
+            avg_z_y.append(avg_z)
+
+        # Calculate corrections to reach the desired point
+        x_corrections = [z_desired - avg for avg in avg_z_x]
+        y_corrections = [z_desired - avg for avg in avg_z_y]
+
+        return x_corrections, y_corrections
+
+    def _start_autocalibration(self, sample_count):
+
+        if not all([
+                self.x_start_point[0],
+                self.x_end_point[0],
+                self.y_start_point[0],
+                self.y_end_point[0]
+                ]):
+                raise self.gcmd.error(
+                    """AXIS_TWIST_COMPENSATION_AUTOCALIBRATE requires
+                    calibrate_start_x, calibrate_end_x, calibrate_start_y
+                    and calibrate_end_y to be defined
+                    """
+                    )
+
+        # check for valid sample_count
+        if sample_count is None or sample_count < 2:
+            raise self.gcmd.error(
+                "SAMPLE_COUNT to probe must be at least 2")
+
+        # verify no other manual probe is in progress
+        manual_probe.verify_no_manual_probe(self.printer)
+
+        # clear the current config
+        self.compensation.clear_compensations()
```

Comment:
This call clears all compensations, but we should remove only compensations for the axis we are calibrating, otherwise we lose one axis each time we calibrate the other.